internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub expr: MonoType,
+}
+
+impl fmt::Display for PolyType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.bnds {
+            Some(bnds) => {
+                let mut bounds = Vec::new();
+                for tv in &self.free {
+                    if let Some(kind) = bnds.get(tv) {
+                        bounds.push(BoundTvar {
+                            tv: *tv,
+                            kind: *kind,
+                        })
+                    }
+                }
+                write!(
+                    f,
+                    "forall [{}] where {} {}",
+                    List { values: &self.free },
+                    List { values: &bounds },
+                    self.expr,
+                )
+            }
+            None => write!(f, "forall [{}] {}", List { values: &self.free }, self.expr),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Add,
+    Div,
+    Cmp,
+    Nul,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Add => f.write_str("Addable"),
+            Kind::Div => f.write_str("Divisible"),
+            Kind::Cmp => f.write_str("Comparable"),
+            Kind::Nul => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Flt,
+    Str,
+    Dur,
+    Time,
+    Rgx,
+    Var(Tvar),
+    Arr(Box<Array>),
+    Rec(Box<Row>),
+    Fun(Box<Function>),
+}
+
+impl fmt::Display for MonoType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MonoType::Bool => f.write_str("bool"),
+            MonoType::Int => f.write_str("int"),
+            MonoType::Uint => f.write_str("uint"),
+            MonoType::Flt => f.write_str("flt"),
+            MonoType::Str => f.write_str("str"),
+            MonoType::Dur => f.write_str("dur"),
+            MonoType::Time => f.write_str("time"),
+            MonoType::Rgx => f.write_str("rgx"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Rec(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
+
+impl fmt::Display for Tvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "t{}", self.0)
```

Comment:
Should this include the single quote to be consistent with the type expression grammar? Meaning lie this `'t0`?

Should we map numeric types to the alphabet? Meaning use `'a`, `'z`, `'aa` and `'ab` for `'t0`, `'t25`, `'t26` and `'t27` respectively? Not sure this is a good idea, just thinking out loud.

Reply:
I would like the display code to be as small and concise as possible. If possible it should be immediately obvious to someone who is reading the code, how types are displayed in my opinion. So my vote would be to not include the single quote or any alphabet mapping.

Reply:
ok, should we change the type expression grammar to not use single quotes? I am operating under the assumption that the displayed types should be parse-able type expressions. That way as developers or users we speak the same language regarding types.

Reply:
Yes the displayed types should be valid type expressions. The single quote form and `t` prefix form should both be parse-able. This should be specified as part of #1952.

Reply:
why do we want both forms? I'll create a separate issue to discuss this fully, I don't want to block this PR on the grammar of the type expressions.

Reply:
Issue https://github.com/influxdata/flux/issues/1993